File: library/python/rolling-hash.py

# Function calculates rolling hash of a string of lowercase english letters
def get_hash(string: str)->int:
    MOD, p, p_power, value = 10**9+7, 31, 1, 0
    for c in string:
        value = (value + (ord(c)-ord('a')+1)*p_power)%MOD
        p_power *= p
    return value
if __name__ == "__main__":
    print(get_hash('ab'))

File: library/cpp/rolling-hash.cpp
#include <bits/stdc++.h>
using namespace std;

long long int get_hash(string);
int main() {
    // Test code
    cout << get_hash("ruts") << endl;
}
/*
    Function calculates polynomial hash of a string of lowercase english letters
*/
long long int get_hash(string str) {
    int p = 31;
    int p_power = 1;
    long long int value = 0;
    for (char c : str) {
        value += (c-'a'+1)*p_power;
        p_power *= p;
    }
    return value;
}

